gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspExtension.kt

```diff
@@ -18,32 +18,146 @@
 package com.google.devtools.ksp.gradle
 
 import org.gradle.api.GradleException
+import org.gradle.api.Project
 import org.gradle.process.CommandLineArgumentProvider
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
 
+/**
+ * A Gradle extension to configure KSP.
+ */
 open class KspExtension {
-    internal val apOptions = mutableMapOf<String, String>()
+    private val apOptions = mutableMapOf<String, String>()
     internal val commandLineArgumentProviders = mutableListOf<CommandLineArgumentProvider>()
 
-    open val arguments: Map<String, String> get() = apOptions.toMap()
+    private val sourceSetOptions = mutableMapOf<String, SourceSetOptions>()
+
+    // Some options have a global and a source set-specific variant. The latter, if specified, overrides the former.
+    // The following is necessary due to KotlinSourceSet not being extension-aware:
+    // - A KotlinSourceSet receiver addresses the source set-specific variant if the `ksp { ... }` block is invoked
+    //   inside a source set block.
+    // - A Project receiver addresses the corresponding global variant. (This is required as the compiler's name
+    //   resolution would always prefer a receiver-less member and never invoke an extension receiver variant.)
+    // - A corresponding private property provides the global variant's backing field.
```

Comment:
Most probably not (as we need to have the upstream change of `KotlinSourceSet` being extension-aware), but I still have this on my radar and will check. It Groovy support a KSP goal? (As KSP is for Kotlin and there is not a single test based on Groovy build scripts, I'm wondering.)

Reply:
We use kotlin script in KSP project itself, but there are lots of end users still uses Groovy, so I would prefer a groovy support, if possible.

Reply:
Actually, with the current PR, Groovy accessors do not work. In fact, it breaks the global `ksp { ... }` block.

A GitHub code search finds
* 40 files using Groovy build scripts – search term `content:/ksp *\{/ path:/\.gradle$/ NOT is:archived` 
* 91 files using Kotlin build scripts – search term `content:/ksp *\{/ path:/\.gradle.kts$/ NOT is:archived` 

To avoid breaking Groovy support, we could either
* enable the new functionality dependent on a Gradle property, say `ksp.multiplatform.enabled`, or
* register a second extension, which is then used exclusively for source sets, e.g. `kspSourceSet { ... }`.

The first alternative would have the advantage of not requiring changes once `KotlinSourceSet` becomes extension aware, so I'd suggest using the property to switch. OK for you?

Reply:
Commit 9cf4738ec63cac385655b82a9f1af97600028485 will stop breaking Groovy build scripts, making the source set configuration dependent on a Gradle property setting `ksp.multiplatform.enabled=true`. Configuring source sets va Groovy will be supported later, once `KotlinSourceSet` becomes extension-aware.